[multistage] Implement ordering for SortExchange

somandal opened the pull request:
This PR adds support for ordering on the collation keys when a `SortExchange` is present in the relational tree. To do this the following changes have been made:

- Add a `PinotLogicalSortExchange` to replace usage of `LogicalSortExchange`. This is needed because later `LogicalSortExchange` is broken into a `MailboxSendNode` and `MailboxReceiveNode`. Information on where to perform ordering (sender, receiver or both) needs to be passed down to these nodes and this is not possible via the `LogicalSortExchange` class alone.
    - We had discussed the potential to make `Exchange` hintable, but based on the hint semantics it should only be applied at the table or `SELECT` level which doesn't apply to `Exchange` nodes. The alternative was to extend the `SortExchange` with our own implementation.
- Add support for ordering based on collation keys in the `MailboxReceiveOperator` when `PinotLogicalSortExchange` is used. 
    - `MailboxSendOperator` will be modified later to add sort support. Once the sender can sort, the receiver needs to be modified to do a k-way merge instead of `PriorityQueue` based sort if the input is sorted.
- Modify `SortOperator` to avoid sorting if the input is already sorted. It should still apply offset + limit

We will also integrate with `SqlHint` support to decide where to sort (sender, receiver, or both) once the path forward has been figured out on how to get hint support for some of the Apache Calcite nodes such as Window, Sort, Filter, etc. `SqlHint` support has not been added to this PR.

This PR is a pre-requisite to adding support for Window Functions with `ORDER BY`.

Design document: https://docs.google.com/document/d/13CmFm4djI09JKF_Xty5acoXxJoxC9CLXAsmikgzgtIs/edit?usp=sharing (discusses the design on ordering options for window functions, one of which is to sort via `LogicalSortExchange`)
Window functions Issue: https://github.com/apache/pinot/issues/7213
Window Function Planner PR: https://github.com/apache/pinot/pull/10228
Window Functions Runtime PR for empty `OVER()` and `OVER(PARTITION BY)`: https://github.com/apache/pinot/pull/10286

cc @walterddr @siddharthteotia @vvivekiyer @ankitsultana

walterddr commented on pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/MailboxReceiveNode.java:
```diff
@@ -43,11 +57,25 @@ public MailboxReceiveNode(int stageId) {
 
   public MailboxReceiveNode(int stageId, DataSchema dataSchema, int senderStageId,
       RelDistribution.Type exchangeType, @Nullable KeySelector<Object[], Object[]> partitionKeySelector,
+      @Nullable List<RelFieldCollation> fieldCollations, boolean isInputSorted, boolean sortOnReceiver,
```
+1 but i dont think this PR would be related since the refactoring only deal with visitor invariant, shared info across the stage such as partition, stageID, etc. 

but i would anticipate a rebase issue. thank you for calling it out @ankitsultana

somandal replied:
+1 this will only help with future shared info. The fields I've added as part of this PR aren't shared.

walterddr commented on pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/MailboxSendNode.java:
```diff
@@ -32,17 +38,36 @@ public class MailboxSendNode extends AbstractStageNode {
     _receiverStageId = receiverStageId;
     _exchangeType = exchangeType;
     _partitionKeySelector = partitionKeySelector;
+    if (!CollectionUtils.isEmpty(fieldCollations)) {
```
can we add a TODO or Preconditions check to indicate that the sorting is not supported?

somandal replied:
done

walterddr reviewed (commented):
the design approach look good to me overall. will take a look at the implementation details next. thank you for the contribution @somandal

walterddr commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MailboxReceiveOperator.java:
```diff
@@ -176,6 +214,18 @@ protected TransferableBlock getNextBlock() {
       }
     }
 
+    if (((openMailboxCount == 0) || (openMailboxCount <= eosMailboxCount))
+        && (!CollectionUtils.isEmpty(_priorityQueue)) && !_isSortedBlockConstructed) {
+      // Some data is present in the PriorityQueue, these need to be sent upstream
+      LinkedList<Object[]> rows = new LinkedList<>();
+      while (_priorityQueue.size() > 0) {
+        Object[] row = _priorityQueue.poll();
+        rows.addFirst(row);
+      }
+      _isSortedBlockConstructed = true;
+      return new TransferableBlock(rows, _dataSchema, DataBlock.Type.ROW);
+    }
```
based on this logic. as long as there's a _priorityQueue and size is not zero you should return the priority queue results. 
when any mailbox receives an error. it should also dropped any _priorityQueue content and stopped the consumption.

somandal replied:
Agreed, I've modified the code to drop the contents of the `_priorityQueue` on receiving an error block. Keeping this block as is. Let me know if you intended for a different change here.

walterddr commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MailboxSendOperator.java:
```diff
@@ -115,6 +125,11 @@ public MailboxSendOperator(MailboxService<TransferableBlock> mailboxService,
+    _collationKeys = collationKeys;
+    _collationDirections = collationDirections;
+    _sortOnSender = sortOnSender;
+    Preconditions.checkState(!sortOnSender, "Sort on send is not yet implemented!");
```
do not check this on the sender operator. only check in the sending node side. the support precondition is a planner constrain not a runtime constrain

somandal replied:
done

walterddr reviewed (commented):
looks good to me mostly. 

i have some concerns regarding how the mailbox (1) handles received block; (2) what should be returned when normal block or error block is being received. please kindly take a look

somandal commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MailboxReceiveOperator.java:
```diff
@@ -161,7 +194,12 @@ protected TransferableBlock getNextBlock() {
               return _upstreamErrorBlock;
             }
             if (!block.isEndOfStreamBlock()) {
-              return block;
+              if (_priorityQueue != null) {
+                List<Object[]> container = block.getContainer();
+                _priorityQueue.addAll(container);
+              } else {
+                return block;
+              }
```
So the code here does return a no-op block (after exiting out of the loop, i.e processing all the mailboxes). Are you recommending instead to return a no-op block for each mailbox? (at least it seems so from the recommended change)

I've created the util as recommended and it explicitly returns a no-op block after adding all the rows to the `PriorityQueue` if the data needs to be ordered for that given mailbox.

Any reason why we should not keep processing all mailboxes and instead return a no-op like recommended?

somandal commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java:
```diff
@@ -213,7 +213,7 @@ private void runLeafStage(DistributedStagePlan distributedStagePlan, Map<String,
           new LeafStageTransferableBlockOperator(serverQueryResults, sendNode.getDataSchema(), requestId,
               sendNode.getStageId(), _rootServer), receivingStageMetadata.getServerInstances(),
-          sendNode.getExchangeType(), sendNode.getPartitionKeySelector(), _rootServer, requestId,
+          sendNode.getExchangeType(), sendNode.getPartitionKeySelector(), null, null, false, _rootServer, requestId,
```
done

somandal commented on pinot-query-planner/src/test/java/org/apache/calcite/rel/rules/PinotSortExchangeCopyRuleTest.java:
```diff
@@ -81,7 +81,8 @@ public void tearDown()
     // Given:
-    SortExchange exchange = LogicalSortExchange.create(_input, RelDistributions.SINGLETON, RelCollations.EMPTY);
+    SortExchange exchange = PinotLogicalSortExchange.create(_input, RelDistributions.SINGLETON, RelCollations.EMPTY,
+        false, true);
```
So testing this is a bit tricky since on the Planner side I don't set any combination except `(false, true)` for now. These tests don't actually create nodes where we fail the creation of the nodes. I was planning to add tests with the hints once we add hint support to ensure that if the hint is set we fail since we don't support ordering on send.

Okay to take this up when hint support is added?

somandal commented on pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/MailboxSendNode.java:
```diff
@@ -32,17 +39,38 @@ public class MailboxSendNode extends AbstractStageNode {
+  @ProtoProperties
+  private List<RelFieldCollation.Direction> _collationDirections;
+  @ProtoProperties
+  private boolean _sortOnSender;
```
done

somandal commented on pinot-query-planner/src/main/java/org/apache/calcite/rel/logical/PinotLogicalSortExchange.java:
```diff
@@ -0,0 +1,111 @@
+
+  public boolean isSortedOnReceiver() {
+    return _isSortedOnReceiver;
+  }
```
done

somandal commented on pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/MailboxReceiveNode.java:
```diff
@@ -32,6 +38,14 @@ public class MailboxReceiveNode extends AbstractStageNode {
+  @ProtoProperties
+  private boolean _isInputSorted;
+  @ProtoProperties
+  private boolean _sortOnReceiver;
```
done

somandal commented on pinot-query-planner/src/main/java/org/apache/calcite/rel/logical/PinotLogicalSortExchange.java:
```diff
@@ -0,0 +1,111 @@
+        newDistribution, newCollation, _isSortedOnSender, _isSortedOnReceiver);
+  }
+
+  @Override public RelWriter explainTerms(RelWriter pw) {
```
done

somandal commented:
> i have some concerns regarding how the mailbox (1) handles received block; (2) what should be returned when normal block or error block is being received. please kindly take a look

Hey @walterddr can you check if I addressed your comments in this area correctly? We can discuss this some more if required.

Basically made changes to:
1) Clear the `_priorityQueue` when an error block is received / returned.
2) Always return a no-op when a block is received and the contents are added to the `_priorityQueue`. Earlier I was waiting for all the mailboxes to get processed in the loop before returning a no-op, now I return a no-op immediately for each mailbox.

walterddr commented:
> > i have some concerns regarding how the mailbox (1) handles received block; (2) what should be returned when normal block or error block is being received. please kindly take a look
> 
> Hey @walterddr can you check if I addressed your comments in this area correctly? We can discuss this some more if required.
> 
> Basically made changes to:
> 
> 1. Clear the `_priorityQueue` when an error block is received / returned.
> 2. Always return a no-op when a block is received and the contents are added to the `_priorityQueue`. Earlier I was waiting for all the mailboxes to get processed in the loop before returning a no-op, now I return a no-op immediately for each mailbox.

Yeah when I thought about it more last night I think we shouldn't return a no-op block to every mailbox received.
The problem I have was to achieve 2 goals:
1. collaborative multi-threading. when some mailboxes (or all mailboxes) are empty, they should give up the thread-worker, so another query stage can run. 
    - This is achieved via the return of the no-op block. 
    - Some operator doesn't follow this. for example, AggregateOperator will hold onto the thread-worker until all inbound blocks are consumed and created an indexed map.
    - this can be problematic as when multiple stages are running on the same server, it could deadlock awaiting inbound messages; but the inbound stage cannot be scheduled due to thread starvation.

2. fairness when grabbing from all mailboxes. 
    - this is more related to the K-loser algorithm we will implement when the sender side is sorting. this way we will maintain some sort of state internally in the sender/receive operator.
    - however, we will actually return some blocks as long as those blocks contain data smaller (or larger when ORDER DESC) than any top row from all mailbox blocks.

for now the implementation looks good. i will create an issue to systematically address both points above

walterddr reviewed (approved):
lgtm
